File: lib/networks/network_context.go
/*
 * Copyright (c) 2020 - present Kurtosis Technologies LLC.
 * All Rights Reserved.
 */

package networks

import (
	"context"
	"fmt"
	"github.com/google/uuid"
	"github.com/kurtosis-tech/kurtosis-go/lib/client/artifact_id_provider"
	"github.com/kurtosis-tech/kurtosis-go/lib/core_api/bindings"
	"github.com/kurtosis-tech/kurtosis-go/lib/kurtosis_service"
	"github.com/kurtosis-tech/kurtosis-go/lib/kurtosis_service/method_types"
	"github.com/kurtosis-tech/kurtosis-go/lib/services"
	"github.com/kurtosis-tech/kurtosis-go/lib/test_suite_docker_consts/test_suite_container_mountpoints"
	"github.com/palantir/stacktrace"
	"github.com/sirupsen/logrus"
	"os"
	"path"
	"path/filepath"
	"sync"
	"time"
)

const (
	// NOTE: This is kinda weird - when we remove a service we can never get it back so having a container
	//  stop timeout doesn't make much sense. It will make more sense when we can stop/start containers
	// Independent of adding/removing them from the network
	removeServiceContainerStopTimeout = 10 * time.Second

	// This will alwyas resolve to the default partition ID (regardless of whether such a partition exists in the network,
	//  or it was repartitioned away)
	defaultPartitionId PartitionID = ""
)

type NetworkContext struct {
	client bindings.TestExecutionServiceClient

	filesArtifactUrls map[services.FilesArtifactID]string
}


/*
Creates a new NetworkContext object with the given parameters.

Args:
	client: The Kurtosis API client that the NetworkContext will use for modifying the state of the testnet
	filesArtifactUrls: The mapping of filesArtifactId -> URL for the artifacts that the testsuite will use
*/
func NewNetworkContext(
		client bindings.TestExecutionServiceClient,
		filesArtifactUrls map[services.FilesArtifactID]string) *NetworkContext {
	return &NetworkContext{
		client: client,
		filesArtifactUrls: filesArtifactUrls,
	}
}

/*
Adds a service to the network in the default partition with the given service ID

NOTE: If the network has been repartitioned and the default partition hasn't been preserved, you should use
	AddServiceToPartition instead.

Args:
	serviceId: The service ID that will be used to identify this node in the network.
	initializer: The Docker container initializer that contains the logic for starting the service

Return:
	service: The new service
*/
func (networkCtx *NetworkContext) AddService(
		serviceId services.ServiceID,
		initializer services.DockerContainerInitializer) (services.Service, services.AvailabilityChecker, error) {
	// Go mutexes aren't re-entrant, so we lock the mutex inside this call
	service, availabilityChecker, err := networkCtx.AddServiceToPartition(
		serviceId,
		defaultPartitionId,
		initializer)
	if err != nil {
		return nil, nil, stacktrace.Propagate(err, "An error occurred adding service '%v' to the network in the default partition", serviceId)
	}
	return service, availabilityChecker, nil
}

/*
Adds a service to the network with the given service ID, created using the given configuration ID.

NOTE: If the network hasn't been repartitioned yet, the PartitionID should be an empty string to add to the default
	partition.

Args:
	serviceId: The service ID that will be used to identify this node in the network.
	partitionId: The partition ID to add the service to
	initializer: The Docker container initializer that contains the logic for starting the service

Return:
	service.Service: The new service
*/
func (networkCtx *NetworkContext) AddServiceToPartition(
		serviceId services.ServiceID,
		partitionId PartitionID,
		initializer services.DockerContainerInitializer) (services.Service, services.AvailabilityChecker, error) {
	ctx := context.Background()

	logrus.Tracef("Registering new service ID with Kurtosis API...")
	registerServiceArgs := &bindings.RegisterServiceArgs{
		ServiceId:       string(serviceId),
		PartitionId:     string(partitionId),
		FilesToGenerate: initializer.GetFilesToMount(),
	}
	registerServiceResp, err := networkCtx.client.RegisterService(ctx, registerServiceArgs)
	if err != nil {
		return nil, nil, stacktrace.Propagate(err, "An error occurred registering service with ID '%v' with the Kurtosis API")
	}
	logrus.Tracef("New service successfully registered with Kurtosis API")

	generatedFilesRelativeFilepaths := registerServiceResp.GeneratedFilesRelativeFilepaths
	generatedFilesFps := map[string]*os.File{}
	generatedFilesAbsoluteFilepaths := map[string]string{}
	for fileId, relativeFilepath := range generatedFilesRelativeFilepaths {
		absoluteFilepath := path.Join(test_suite_container_mountpoints.SuiteExVolMountpoint, relativeFilepath)
		generatedFilesAbsoluteFilepaths[fileId] = absoluteFilepath
		fp, err := os.Create(absoluteFilepath)
		if err != nil {
			return nil, nil, stacktrace.Propagate(
				err,
				"Could not open generated file '%v' for writing",
				fileId)
		}
		defer fp.Close()
		generatedFilesFps[fileId] = fp
	}

	logrus.Trace("Initializing generated files...")
	if err := initializer.InitializeMountedFiles(generatedFilesFps); err != nil {
		return nil, nil, stacktrace.Propagate(err, "An error occurred initializing the generated files")
	}
	logrus.Trace("Successfully initialized generated files")


	logrus.Tracef("Creating files artifact mount dirpaths map...")
	filesArtifactMountDirpaths := map[string]string{}
	for filesArtifactId, mountDirpath := range initializer.GetFilesArtifactMountpoints() {
		filesArtifactMountDirpaths[string(filesArtifactId)] = mountDirpath
	}
	logrus.Tracef("Successfully created files artifact mount dirpaths map")

	logrus.Tracef("Creating start command for service...")
	serviceIpAddr := registerServiceResp.IpAddr
	startCmdArgs, err := initializer.GetStartCommand(generatedFilesAbsoluteFilepaths, serviceIpAddr)
	if err != nil {
		return nil, nil, stacktrace.Propagate(err, "Failed to create start command")
	}
	logrus.Tracef("Successfully created start command for service")

	logrus.Tracef("Starting new service with Kurtosis API...")
	dockerImage := initializer.GetDockerImage()
	startServiceArgs := &bindings.StartServiceArgs{
		ServiceId:                   string(serviceId),
		DockerImage:                 initializer.GetDockerImage(),
		UsedPorts:                   initializer.GetUsedPorts(),
		StartCmdArgs:                startCmdArgs,
		DockerEnvVars:               map[string]string{}, // TODO actually support Docker env vars!
		SuiteExecutionVolMntDirpath: initializer.GetTestVolumeMountpoint(),
		FilesArtifactMountDirpaths:  filesArtifactMountDirpaths,
	}
	if _, err := networkCtx.client.StartService(ctx, startServiceArgs); err != nil {
		return nil, nil, stacktrace.Propagate(err, "An error occurred starting the service with the Kurtosis API")
	}
	logrus.Tracef("Successfully started service with Kurtosis API")

	logrus.Tracef("Getting service from IP...")
	service := initializer.GetService(serviceIpAddr)
	logrus.Tracef("Successfully got service from IP")

	availabilityChecker := services.NewDefaultAvailabilityChecker(serviceId, service)

	return service, availabilityChecker, nil
}

/*
Stops the container with the given service ID, and removes it from the network.
*/
func (networkCtx *NetworkContext) RemoveService(serviceId services.ServiceID, containerStopTimeoutSeconds int) error {
	logrus.Debugf("Removing service '%v'...", serviceId)
	args := &bindings.RemoveServiceArgs{
		ServiceId:                   string(serviceId),
		ContainerStopTimeoutSeconds: removeServiceContainerStopTimeout,
	}
	if _, err := networkCtx.client.RemoveService(context.Background(), args); err != nil {
		return stacktrace.Propagate(err, "An error occurred removing service '%v' from the network", serviceId)
	}
	logrus.Debugf("Successfully removed service ID %v", serviceId)
	return nil
}

/*
Constructs a new repartitioner builder in preparation for a repartition.

Args:
	isDefaultPartitionConnectionBlocked: If true, when the connection details between two partitions aren't specified
		during a repartition then traffic between them will be blocked by default
 */
func (networkCtx NetworkContext) GetRepartitionerBuilder(isDefaultPartitionConnectionBlocked bool) *RepartitionerBuilder {
	// This function doesn't need a mutex lock because (as of 2020-12-28) it doesn't touch internal state whatsoever
	return newRepartitionerBuilder(isDefaultPartitionConnectionBlocked)
}

/*
Repartitions the network using the given repartitioner. A repartitioner builder can be constructed using the
	NewRepartitionerBuilder method of this network context object.
 */
func (networkCtx *NetworkContext) RepartitionNetwork(repartitioner *Repartitioner) error {
	partitionServices := map[string]*bindings.PartitionServices{}
	for partitionId, serviceIdSet := range repartitioner.partitionServices {
		serviceIdStrPseudoSet := map[string]bool{}
		for _, serviceId := range serviceIdSet.getElems() {
			serviceIdStr := string(serviceId)
			serviceIdStrPseudoSet[serviceIdStr] = true
		}
		partitionIdStr := string(partitionId)
		partitionServices[partitionIdStr] = serviceIdStrPseudoSet
	}

	serializablePartConns := map[string]map[string]method_types.SerializablePartitionConnection{}
	for partitionAId, partitionAConns := range repartitioner.partitionConnections {
		serializablePartAConns := map[string]method_types.SerializablePartitionConnection{}
		for partitionBId, unserializableConn := range partitionAConns {
			partitionBIdStr := string(partitionBId)
			serializableConn := makePartConnSerializable(unserializableConn)
			serializablePartAConns[partitionBIdStr] = serializableConn
		}
		partitionAIdStr := string(partitionAId)
		serializablePartConns[partitionAIdStr] = serializablePartAConns
	}

	serializableDefaultConn := makePartConnSerializable(repartitioner.defaultConnection)
	
	repartitionArgs := &bindings.RepartitionArgs{
		PartitionServices:    partitionServices,
		PartitionConnections: serializablePartConns,
		DefaultConnection:    serializableDefaultConn,
	}

	if err := networkCtx.kurtosisService.Repartition(partitionServices, serializablePartConns, serializableDefaultConn); err != nil {
		return stacktrace.Propagate(err, "An error occurred repartitioning the test network")
	}
	return nil
}

// ============================================================================================
//                                    Private helper methods
// ============================================================================================
func makePartConnSerializable(connection PartitionConnection) method_types.SerializablePartitionConnection {
	return method_types.SerializablePartitionConnection{
		IsBlocked: connection.IsBlocked,
	}
}

File: rust/lib/src/networks/network_context.rs
use std::{collections::{HashMap, HashSet}, fs::File, ops::{Deref, DerefMut}, path::{PathBuf}, rc::Rc};
use std::hash::Hash;

use dashmap::DashMap;
use log::{debug, trace};
use tokio::runtime::Runtime;
use tonic::{transport::Channel};
use anyhow::{anyhow, Context, Result};

use crate::{core_api_bindings::api_container_api::{PartitionConnectionInfo, PartitionConnections, PartitionServices, RegisterServiceArgs, RemoveServiceArgs, RepartitionArgs, StartServiceArgs, test_execution_service_client::TestExecutionServiceClient}, services::{availability_checker::AvailabilityChecker, container_config_factory::ContainerConfigFactory, service::{Service, ServiceId}, service_context::ServiceContext}};

use super::network::Network;

// TODO Make a type
const DEFAULT_PARTITION_ID_STR: &str = "";

// This value - where the suite execution volume will be mounted on the testsuite container - is
//  hardcoded inside Kurtosis Core
const SUITE_EX_VOL_MOUNTPOINT: &str = "/suite-execution";

pub type PartitionId = String;

// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
pub struct NetworkContext {
	async_runtime: Rc<Runtime>,
    client: TestExecutionServiceClient<Channel>,
	// TODO Make key a separate FilesArtifactID type
	files_artifact_urls: HashMap<String, String>,
    all_services: DashMap<ServiceId, Rc<dyn Service>>,
}

impl NetworkContext {
    pub fn new(async_runtime: Rc<Runtime>, client: TestExecutionServiceClient<Channel>, files_artifact_urls: HashMap<String, String>) -> NetworkContext {
        return NetworkContext {
			async_runtime,
            client,
			files_artifact_urls,
            all_services: DashMap::new(),
        };
    }

	// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn add_service<S: Service>(&mut self, service_id: &ServiceId, config_factory: &dyn ContainerConfigFactory<S>) -> Result<(Rc<S>, AvailabilityChecker)> {
		let (service_ptr, availability_checker) = self.add_service_to_partition(service_id, &DEFAULT_PARTITION_ID_STR.to_owned(), config_factory)
			.context(format!("An error occurred adding service '{}' to the network in the default partition", service_id))?;
		return Ok((service_ptr, availability_checker));
	}

	// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn add_service_to_partition<S: Service>(&mut self, service_id: &ServiceId, partition_id: &PartitionId, config_factory: &dyn ContainerConfigFactory<S>) -> Result<(Rc<S>, AvailabilityChecker)> {
		trace!("Registering new service ID with Kurtosis API...");
		let args = RegisterServiceArgs{
		    service_id: service_id.to_owned(),
		    partition_id: partition_id.to_owned(),
		};
		let register_service_args = tonic::Request::new(args);
		let register_service_resp = self.async_runtime.block_on(self.client.register_service(register_service_args))
			.context(format!("An error occurred registering service with ID '{}' with the Kurtosis API", service_id))?
			.into_inner();
		let service_ip_addr = register_service_resp.ip_addr;
		let container_creation_config = config_factory.get_creation_config(&service_ip_addr)
			.context("An error occurred getting the container creation config")?;
		let service_context_client = self.client.clone();
		let service_context = ServiceContext::new(
			self.async_runtime.clone(), 
			service_context_client, 
			service_id.to_owned(), 
			service_ip_addr.to_owned(),
			SUITE_EX_VOL_MOUNTPOINT.to_owned(),
			container_creation_config.get_test_volume_mountpoint().to_owned(),
		);
		trace!("New service successfully registered with Kurtosis API");

		trace!("Initializing generated files in suite execution volume...");
		let mut files_to_generate: HashSet<String> = HashSet::new();
		for (file_id, _) in container_creation_config.get_file_generating_funcs() {
			files_to_generate.insert(file_id.to_owned());
		}
		let generated_file_filepaths = service_context.generate_files(files_to_generate)
			.context(format!("An error occurred generating the files needed for service startup"))?;
		let mut generated_files_abs_filepaths_on_service: HashMap<String, PathBuf> = HashMap::new();
		for (file_id, initializing_func_arc_mutex) in container_creation_config.get_file_generating_funcs() {
			let filepaths = generated_file_filepaths.get(file_id)
				.context(format!(
					"Needed to initialize file for file ID '{}', but no generated file filepaths were found for that file ID; this is a Kurtosis bug",
					file_id,
				))?;
			let fp = File::create(&filepaths.absolute_filepath_on_testsuite_container)
				.context(format!("An error occurred opening file pointer for file '{}'", file_id))?;
			let mut initializing_func = initializing_func_arc_mutex.lock().unwrap();
			initializing_func.deref_mut()(fp)
				.context(format!("The function to initialize file with ID '{}' returned an error", file_id))?;
			generated_files_abs_filepaths_on_service.insert(file_id.to_owned(), filepaths.absolute_filepath_on_service_container.clone());
		}
		trace!("Successfully initialized generated files in suite execution volume");

		let container_run_config = config_factory.get_run_config(&service_ip_addr, generated_files_abs_filepaths_on_service)
			.context("An error occurred getting the container run config")?;

		trace!("Creating files artifact URL -> mount dirpaths map...");
		let mut artifact_url_to_mount_dirpath: HashMap<String, String> = HashMap::new();
		for (files_artifact_id, mount_dirpath) in container_creation_config.get_files_artifact_mountpoints() {
			let artifact_url = self.files_artifact_urls.get(files_artifact_id)
				.context(format!(
					"Service requested file artifact '{}', but the network context doesn't have a URL for that file artifact; this is a bug with Kurtosis itself",
					files_artifact_id
				))?;
			artifact_url_to_mount_dirpath.insert(artifact_url.to_owned(), mount_dirpath.to_owned());
		}

		trace!("Successfully created files artifact URL -> mount dirpaths map");

		trace!("Starting new service with Kurtosis API...");
		let start_service_args = StartServiceArgs{
		    service_id: service_id.to_owned(),
		    docker_image: container_creation_config.get_image().to_owned(),
		    used_ports: NetworkContext::convert_hashset_to_hashmap(container_creation_config.get_used_ports().to_owned()),
			entrypoint_args: container_run_config.get_entrypoint_override_args().to_owned(),
			cmd_args: container_run_config.get_cmd_override_args().to_owned(),
		    docker_env_vars: container_run_config.get_environment_variable_overrides().to_owned(),
		    suite_execution_vol_mnt_dirpath: container_creation_config.get_test_volume_mountpoint().to_owned(),
		    files_artifact_mount_dirpaths: artifact_url_to_mount_dirpath,
		};
		let start_service_req = tonic::Request::new(start_service_args);
		self.async_runtime.block_on(self.client.start_service(start_service_req))
			.context("An error occurred starting the service with the Kurtosis API")?
			.into_inner();
		trace!("Successfully started service with Kurtosis API");


		trace!("Creating service interface...");
		let result_service_ptr = container_creation_config.get_service_creating_func()(service_context);
		let casted_result_service_rc = Rc::new(result_service_ptr);
		trace!("Successfully created service interface");

		self.all_services.insert(service_id.to_owned(), casted_result_service_rc.clone());

		let availability_checker = AvailabilityChecker::new(service_id, casted_result_service_rc.clone());

		return Ok((casted_result_service_rc, availability_checker));
    }

	// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn get_service<S: Service>(&self, service_id: &ServiceId) -> Result<Rc<S>> {
		let service_ptr_ptr = self.all_services.get(service_id)
			.context(format!("No service found with ID '{}'", service_id))?;
		let service_ptr = service_ptr_ptr.deref().clone();
		let casted_service_ptr_or_err = service_ptr.downcast_rc::<S>();
		let result: Rc<S>;
		match casted_service_ptr_or_err {
			Ok(casted_ptr) => result = casted_ptr,
			Err(_) => return Err(anyhow!(
				"An error occurred casting service with ID '{}' to appropriate type",
				service_id,
			)),
		}
		return Ok(result);
    }

	// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn remove_service(&mut self, service_id: &ServiceId, container_stop_timeout_seconds: u64) -> Result<()> {
		debug!("Removing service '{}'...", service_id);
		let args = RemoveServiceArgs{
		    service_id: service_id.to_owned(),
			// NOTE: This is kinda weird - when we remove a service we can never get it back so having a container
			//  stop timeout doesn't make much sense. It will make more sense when we can stop/start containers
			// Independent of adding/removing them from the network
		    container_stop_timeout_seconds,
		};
		let req = tonic::Request::new(args);
		self.async_runtime.block_on(self.client.remove_service(req))
			.context(format!("An error occurred removing service '{}' from the network", service_id))?;
		self.all_services.remove(service_id);
		debug!("Successfully removed service ID '{}'", service_id);
		return Ok(());
	}

	// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn repartition_network(
		&mut self, 
		partition_services: HashMap<PartitionId, HashSet<ServiceId>>,
		partition_connections: HashMap<PartitionId, HashMap<PartitionId, PartitionConnectionInfo>>,
		default_connection_info: PartitionConnectionInfo,
	) -> Result<()> {
		let mut req_partition_services: HashMap<String, PartitionServices> = HashMap::new();
		for (partition_id, service_id_set) in partition_services {
			let mut service_id_str_pseudo_set: HashMap<String, bool> = HashMap::new();
			for service_id in service_id_set {
				service_id_str_pseudo_set.insert(service_id, true);
			}
			req_partition_services.insert(partition_id, PartitionServices{
			    service_id_set: service_id_str_pseudo_set,
			});
		}

		let mut req_partition_connections: HashMap<String, PartitionConnections> = HashMap::new();
		for (partition_a_id, partition_a_conns_map) in partition_connections {
			let mut partition_a_conns_str_map: HashMap<String, PartitionConnectionInfo> = HashMap::new();
			for (partition_b_id, conn_info) in partition_a_conns_map {
				partition_a_conns_str_map.insert(partition_b_id, conn_info);
			}
			let partition_a_conns = PartitionConnections{
			    connection_info: partition_a_conns_str_map,
			};
			req_partition_connections.insert(partition_a_id, partition_a_conns);
		}

		let args = RepartitionArgs{
		    partition_services: req_partition_services,
		    partition_connections: req_partition_connections,
			// NOTE: It's unclear why tonic generates this as "optional"; it's not optional in the .proto file
		    default_connection: Some(default_connection_info),
		};

		let req = tonic::Request::new(args);
		self.async_runtime.block_on(self.client.repartition(req))
			.context("An error occurred repartitioning the test network")?;
		return Ok(());
    }

	fn convert_hashset_to_hashmap<T>(
		set: HashSet<T>
	) -> HashMap<T, bool> where T: Eq + Hash {
		let mut result: HashMap<T, bool> = HashMap::new();
		for elem in set {
			result.insert(elem, true);
		}
		return result;
	}
}

impl Network for NetworkContext { }